drape/glyph_manager.cpp

```diff
@@ -659,46 +626,68 @@ hb_language_t OrganicMapsLanguageToHarfbuzzLanguage(int8_t lang)
   hb_language_t const hbLanguage = OrganicMapsLanguageToHarfbuzzLanguage(lang);
 
   text::TextMetrics allGlyphs;
 
-  // TODO(AB): Cache runs.
-  // TODO(AB): Cache buffer.
-  hb_buffer_t * buf = hb_buffer_create();
   for (auto const & substring : segments)
   {
-    hb_buffer_clear_contents(buf);
+    hb_buffer_clear_contents(m_impl->m_harfbuzzBuffer);
 
     // TODO(AB): Some substrings use different fonts.
-    hb_buffer_add_utf16(buf, reinterpret_cast<const uint16_t *>(text.data()), static_cast<int>(text.size()),
-                        substring.m_start, substring.m_length);
-    hb_buffer_set_direction(buf, substring.m_direction);
-    hb_buffer_set_script(buf, substring.m_script);
-    hb_buffer_set_language(buf, hbLanguage);
-
-    // TODO(AB): Check not only the first character to determine font for the run, but all chars.
-    auto firstCharacterIter{text.begin() + substring.m_start};
-    auto const firstCharacterUnicode = utf8::unchecked::next16(firstCharacterIter);
-
-    int const fontIndex = GetFontIndex(firstCharacterUnicode);
-    if (fontIndex < 0)
+    hb_buffer_add_utf16(m_impl->m_harfbuzzBuffer, reinterpret_cast<const uint16_t *>(text.data()),
+        static_cast<int>(text.size()), substring.m_start, substring.m_length);
+    hb_buffer_set_direction(m_impl->m_harfbuzzBuffer, substring.m_direction);
+    hb_buffer_set_script(m_impl->m_harfbuzzBuffer, substring.m_script);
+    hb_buffer_set_language(m_impl->m_harfbuzzBuffer, hbLanguage);
+
+    auto u32CharacterIter{text.begin() + substring.m_start};
+    auto const end{text.begin() + substring.m_start + substring.m_length};
+    do
     {
-      // TODO(AB): Add missing glyph character's metrics
-      LOG(LWARNING, ("Skip run because no font was found for character", NumToHex(firstCharacterUnicode)));
-      continue;
-    }
+      auto const u32Character = utf8::unchecked::next16(u32CharacterIter);
+      // TODO(AB): Mapping characters to fonts can be optimized.
+      int const fontIndex = GetFontIndex(u32Character);
+      if (fontIndex < 0)
+        LOG(LWARNING, ("No font was found for character", NumToHex(u32Character)));
+      else
+      {
+        // TODO(AB): Mapping font only by the first character in a string may fail in theory in some cases.
+        m_impl->m_fonts[fontIndex]->Shape(m_impl->m_harfbuzzBuffer, fontPixelHeight, fontIndex, allGlyphs);
+        break;
+      }
+    } while (u32CharacterIter != end);
+  }
 
-    m_impl->m_fonts[fontIndex]->Shape(buf, fontPixelHeight, fontIndex, allGlyphs);
+  if (allGlyphs.m_glyphs.empty())
+    LOG(LWARNING, ("No glyphs were found in all fonts for string", utf8));
+
+  // Empirically measured, may need more tuning.
+  size_t constexpr kMaxCacheSize = 50000;
+  if (m_impl->m_textMetricsCache.size() > kMaxCacheSize)
+  {
+    LOG(LINFO, ("Clearing text metrics cache"));
+    // TODO(AB): Is there a better way? E.g. clear a half of the cache?
+    m_impl->m_textMetricsCache.clear();
```

Comment:
LRU cache.

Reply:
Won't the overhead from LRU be greater than this simple approach? Browsing some local map area won't fill the cache for sure, and even if the cache is reset, it shouldn't happen often.